src/SignalR/server/StackExchangeRedis/src/RedisHubLifetimeManager.cs

```diff
@@ -476,6 +555,42 @@ private async Task SubscribeToConnection(HubConnectionContext connection)
         channel.OnMessage(channelMessage =>
         {
             var invocation = RedisProtocol.ReadInvocation((byte[])channelMessage.Message);
+
+            // This is a Client result we need to setup state for the completion and send the message to the client
+            if (!string.IsNullOrEmpty(invocation.InvocationId))
+            {
+                CancellationTokenRegistration? tokenRegistration = null;
+                _clientResultsManager.AddInvocation(invocation.InvocationId,
+                    (typeof(RawResult), connection.ConnectionId, null!, (_, completionMessage) =>
+                {
+                    var protocolName = connection.Protocol.Name;
+                    tokenRegistration?.Dispose();
+                    // TODO: acquiring this and then calling RedisProtocol.WriteCompletionMessage will allocate a new MemoryBufferWriter, we can avoid this
+                    var memoryBufferWriter = AspNetCore.Internal.MemoryBufferWriter.Get();
+                    try
+                    {
+                        connection.Protocol.WriteMessage(completionMessage, memoryBufferWriter);
+                        // TODO: we can avoid this ToArray call
+                        var message = RedisProtocol.WriteCompletionMessage(new ReadOnlySequence<byte>(memoryBufferWriter.ToArray()), protocolName);
+                        return PublishAsync(invocation.ReturnChannel!, message);
+                    }
+                    finally
+                    {
+                        memoryBufferWriter.Dispose();
+                    }
+                }
+                ));
+
+                // TODO: this isn't great
```

Comment:
So this races with `tokenRegistration?.Dispose();` right? Is there any impact other than not cleaning up the registration as soon as we could?

Reply:
I can't think of any issues. My biggest issue with this is the closure allocation.